hpct-ldap-server/src/servers/ldap.py

```diff
@@ -0,0 +1,503 @@
+                if len(d) < 1:
+                    raise Exception("Invalid domain components given, cannot be empty.")
+                else:
+                    split_dc.append(f"dc={d}")
+            return ",".join(split_dc)
+        else:
+            return f"dc={dc}"
+
+    def add_user(
+        self,
+        admin_passwd=None,
+        domain=None,
+        gecos=None,
+        gid=None,
+        homedir=None,
+        login=None,
+        passwd=None,
+        uid=None,
+        user=None,
+    ) -> None:
+        """Add user.
+
+        Parameters
+        ----------
+        admin_passwd : str
+            LDAP password.
+        domain : str
+            Domain name.
+        gecos : str
+            Real name.
+        gid : int
+            Group id.
+        homedir : str
+            Home directory path.
+        login : str
+            Login shell.
+        passwd : str
+            User password.
+        uid : int
+            User id.
+        user : str
+            Username.
+        """
+        dcs = self._split_domain(domain)
+
+        if (
+            None not in [admin_passwd, dcs, gecos, gid, homedir, login, passwd, uid, user]
+            and uid > MIN_UID
+            and gid > MIN_GID
+        ):
+            binddn = f"cn=admin,{dcs}"
+            base_user = [
+                f"dn: uid={user.lower()},ou=people,{dcs}",
+                "objectClass: inetOrgPerson",
+                "objectClass: posixAccount",
+                "objectClass: shadowAccount",
+                f"cn: {user.lower()}",
+                f"sn: {user}",
+                f"userPassword: {passwd}",
+                f"loginShell: {login}",
+                f"uidNumber: {uid}",
+                f"gidNumber: {gid}",
+                f"homeDirectory: {homedir}",
+                f"gecos: {gecos}",
+            ]
+
+            with tempfile.NamedTemporaryFile(mode="w+t", delete=True) as f:
+                for line in base_user:
+                    f.write(f"{line}\n")
+                    f.flush()
+                cmd = ["ldapadd", "-x", "-D", binddn, "-w", admin_passwd, "-f", f.name]
+
+                rc = subprocess.call(cmd, stdout=subprocess.DEVNULL, stderr=subprocess.STDOUT)
+                if rc != 0:
+                    raise Exception(
+                        f'Unable to add user. Make sure to run "ldapadd -x -D cn=admin,{dcs} -W -f /etc/ldap/basedn.ldif" first.'
+                    )
+
+    def add_group(
+        self,
+        admin_passwd=None,
+        domain=None,
+        gid=None,
+        group=None,
+    ) -> None:
+        """Add group.
+
+        Parameters
+        ----------
+        admin_passwd : str
+            LDAP password.
+        domain : str
+            Domain name.
+        gid : int
+            Group id.
+        group : str
+            Group name.
+        """
+        dcs = self._split_domain(domain)
+
+        if None not in [domain, gid, group, admin_passwd] and gid > MIN_GID:
```

Comment:
As above, a check for an unacceptable value could allow an early return (with error?) and 1 less indent level below.

Reply:
Updated as suggested